### PR TITLE
HIVE-24388: Enhance swo optimizations to merge EventOperators

### PR DESCRIPTION
**Pull request by kgyrtkirk**

### CODE REVIEW

**Comment by kgyrtkirk** [comment]
@jcamachor could you please take a look? 
this patch also closes down 2 smaller bugs - one of them might have potentially caused parallel edge creation; I think the issue was not new; the new algorithm just stressed the issue a little bit more...I will ask Sungwoo to check it again after we have this patch

File: ql/src/test/results/clientpositive/llap/swo_event_merge.q.out
Comment (jcamachor) [inline ql/src/test/results/clientpositive/llap/swo_event_merge.q.out:199]: Could merging these TS operators be effectively worse?

For instance, in this specific mock query, no partition will be pruned for the *x1_store_sales* table, while before partition pruning was kicking in. Thus, in this case, you are scanning the same data whether you have one or two TS operators (two partitions), however after merging the TS, the size of the data you are shuffling for the join doubles (data in both partitions twice)? Is that analysis correct?

Off the top of my head, this could be beneficial if i) both TS only select a small subset of the partitions in the table, or ii) overlapping in the partition list for those two different TS is greater than a certain threshold. Should we work in that direction, i.e., introduce some config parameters for this?

@rbalamohan , what is your take? It would be helpful to have a second opinion.

Comment (kgyrtkirk) [reply ql/src/test/results/clientpositive/llap/swo_event_merge.q.out:199]: I think right now we don't have a sanity check filter right before the TS to limit reduce shuffled data size to the previous amount; but reducing the number of scans is beneficial - IIRC in the q23 query the scanned partitions were the same; so the benefit was real.

If we make available these things the same way as we have the SJ filters - then we could for sure avoid shuffling more data.

Note: I think it would enable some further opportunities if we would change the SJ data transmission method from RS to EVENTOP - that way we shouldn't have to worry about parallel edges anymore...

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java
Comment (kgyrtkirk) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java:20]: yeah...I've some stuff which write out `File`-s for now...and I just remove it at the end of preparing the patch....
it gives easier insights into what happened in the SWO.

I would really like to build it into the system and expose it on the HS2 web interface - it would be very usefull...will get to that as well :)
but first I would like to finish this set of patches :)

File: ql/src/test/results/clientpositive/llap/swo_event_merge.q.out
Comment (kgyrtkirk) [reply ql/src/test/results/clientpositive/llap/swo_event_merge.q.out:199]: @jcamachor should I add a config knob to control event operator merge?

Comment (rbalamohan) [reply ql/src/test/results/clientpositive/llap/swo_event_merge.q.out:199]: >> however after merging the TS, the size of the data you are shuffling for the join doubles


Yes, this is possible with merging. +1 on adding an option for enabling it. 

There is additional option to reduce number of group by operator evaluations (e.g [Q65](https://github.com/apache/hive/blob/master/ql/src/test/queries/clientpositive/perf/query65.q#L18) computes group by operator twice) which can help in reducing runtime.

**Review by jcamachor — APPROVED** [review]

**Pull request merged**
